server/functions/array_to_string.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// initArrayToString registers the functions to the catalog.
+func initArrayToString() {
+	framework.RegisterFunction(array_to_string2)
+	framework.RegisterFunction(array_to_string3)
+}
+
+// array_to_string represents the PostgreSQL system catalog information function taking 2 parameters.
+var array_to_string2 = framework.Function2{
+	Name:               "array_to_string",
+	Return:             pgtypes.Text,
+	Parameters:         [2]pgtypes.DoltgresType{pgtypes.AnyArray, pgtypes.Text},
+	IsNonDeterministic: true,
+	Callable: func(ctx *sql.Context, _ [3]pgtypes.DoltgresType, val1, val2 any) (any, error) {
+		arr := val1.([]any)
+		delimiter := val2.(string)
+		strs := make([]string, 0)
+		for _, el := range arr {
+			if el != nil {
+				strs = append(strs, fmt.Sprintf("%v", el))
```

Comment:
We should almost never use `fmt.Sprintf("%v")` for converting values to their text counterparts. We should always follow what Postgres does for these functions, and Postgres is going to either use the `IoOutput` function or the `text` cast for the type.

We can test this in Postgres by creating our own type based on another type. Our new type will inherit the I/O functions, which is useful for our check.
```sql
CREATE TYPE test_float8 AS (value float8);
```
Then we create the function that will be called for our cast. In this example that I'm using, we make the cast so that it returns a string containing what's after the decimal point. That makes it very easy to see whether Postgres is using the cast or the I/O functions, since we know the I/O functions will include the full number. We'll also create a function so that we can convert `numeric` types to our new type for ease of use:
```sql
CREATE FUNCTION test_float8_to_text(test_float8) RETURNS text AS $$
    DECLARE
        result text;
    BEGIN
        result := MOD($1.value::numeric, 1);
        RETURN result;
    END;
$$ LANGUAGE plpgsql IMMUTABLE STRICT;

CREATE FUNCTION numeric_to_test_float8(numeric) RETURNS test_float8 AS $$
    DECLARE
        result test_float8;
    BEGIN
        result.value := $1::float8;
        RETURN result;
    END;
$$ LANGUAGE plpgsql IMMUTABLE STRICT;
```
Then we can create the casts:
```sql
CREATE CAST (test_float8 AS text) WITH FUNCTION test_float8_to_text(test_float8) AS IMPLICIT;
CREATE CAST (numeric AS test_float8) WITH FUNCTION numeric_to_test_float8(numeric) AS IMPLICIT;
```
Finally, we can query our `array_to_string` function to see how it's working. We wouldn't have to do all of this if the documentation stated that it used either a cast or the I/O functions, but since it doesn't, we have to manually test it to find out.
```sql
SELECT array_to_string(ARRAY[37.89::test_float8, 1.2::test_float8], '_');
```
This returns `(37.89)_(1.2)`. We can ignore the parenthesis as they're due to the I/O function in use (`record_out`), but we can see that it's **not** using the cast function, since it would have `0.89` and `0.2` instead. Also, for a sanity check, we can use our cast to make sure that it's returning as we expect.
```sql
SELECT 37.89::test_float8::text;
```
This returns `"0.89"`, so we can confirm that it's using the I/O output function.

This might seem like a lot, but this all becomes very straightforward to check once you're familiar with the type system. Generally it may be a bit more obvious on how the function works (as there may only be one reasonable way), but in this case there are two possible paths and the documentation does not provide any clues, so we have to find out for ourselves.